### DISCUSSION TITLE
What made you choose Rust 1.67.0?

### DISCUSSION THREAD

**Discussion by steveklabnik** [fish-shell/fish-shell#10230] [General]
I don't think this is a bad thing or a good thing, inherently, but I couldn't find any documentation on why this version was chosen, and I think knowing the why would be interesting!

**Comment by zanchey** [comment]
The widestring_suffix proc macro needed it, but we've dropped that, so could possibly look at lowering it.

**Reply by zanchey** [reply]
1.64 looks like the minimum while we're still using [corrosion](https://github.com/corrosion-rs/corrosion).

**Reply by krobelus** [reply]
I think they require 1.46, not 1.64

**Reply by steveklabnik** [reply]
Honestly, I am surprised by how low it is, not suggesting you should drop any lower!

maybe a good rephrase is, "is there any criteria you all are using for minimum rust versions?" I know some projects choose things like "latest stable - 3", others would choose something like "the version in current debian stable."

**Reply by zanchey** [reply]
Sorry, the error I thought was from Corrosion was actually from Cargo.

Workspace inheritance requires 1.64
std::backtrace and let-else needs 1.65
char::methods::<impl char>::from_u32 and the panic!() macro, as well as some of the pattern matching we use, needs 1.66

1.67 was the latest stable when the RIIR branch was merged. I don't think we'll bump it unless it really benefits us; see @faho's comment here which I concur with: https://github.com/fish-shell/fish-shell/pull/9944#discussion_r1291315610

**Reply by faho** [reply]
> is there any criteria you all are using for minimum rust versions?

Just simply "what we need". We would go higher if there's a compelling reason for it, and so far we haven't had a lot.

E.g. like zanchey linked we could have increased MSRV for [is_terminal](https://docs.rs/is-terminal/latest/is_terminal/), but then I saw that it literally just calls isatty on unix (and we're not interested in windows), so there's no real point to it.

Basically: We're used to the pain of not requiring C++14 in 2021 because that would be too odious for old systems, by comparison rust 1.67 is nothing. (we have a few lint-allows we could remove and some small backports for is_some_and and friends)

And there would be value in making it build with e.g. Debian's distro rustc (currently 1.63, next is 1.70) if we can achieve that - this isn't for rust nerds but geared towards classic unix types.

> I know some projects choose things like "latest stable - 3", others would choose something like "the version in current debian stable."

I'm really not a friend of just automatically bumping the version. Our policy is close-ish to "current debian stable", except we started with 1.67 so we're at the *next* debian stable right now.

Also I'm okay with bumping it past that *if* there is a compelling need. We're pretty new to this and so it's not clear to me how workable tying to current debian stable is.

So I guess you could summarize it as "be modest, let's see if we can get by with what's in Debian Stable", basically

**Reply by steveklabnik** [reply]
Thanks! Appreciate the context.

**Reply by krobelus** [reply]
debian stable has had a C++20 compiler for a while now, I think one of the problems was systems like centos 7

**Comment by faho** [comment]
We've increased the MSRV to 1.70 now because no distro ships 1.67, 1.68 or 1.69.

So it wasn't useful to stay there and we didn't have the appetite to *decrease* the MSRV to Debian Stable's 1.63. 1.70 is the *next* Debian Stable tho.

**Comment by g2p** [comment]
Here's a note that 1.82 would be useful for &raw access to struct members:
https://blog.rust-lang.org/2024/10/17/Rust-1.82.0.html#native-syntax-for-creating-a-raw-pointer
https://github.com/fish-shell/fish-shell/issues/11221

**Reply by steveklabnik** [reply]
I'm guessing 1.85 will end up being a common MSRV for the ecosystem in the future, with the 2024 edition and stuff like async closures.

But yeah, `&raw` is really nice.

**Comment by ridiculousfish** [comment]
We use 1.73.0 as this was the last version to support macOS 10.9; we're likely to drop support for that in the next major version bump.

**Comment by ghost** [comment]
I embrace every version of Rust, all because of Fish shell — my favorite. Fish opened the gateway to Rust, and now I truly rust within Fish. May Fish always swim happily through the digital ocean.
